velox/functions/lib/LikeFunctionUtils.cpp

```diff
@@ -0,0 +1,215 @@
+// the longest proper prefix that is also a suffix of the pat[0-i].
+void kmpSearchPreprocessPattern(
+    const std::string& pattern,
+    std::vector<int>& lps) {
+  size_t patternIndex = 1;
+  size_t lpsLength = 0;
+  size_t patternLength = pattern.size();
+
+  while (patternIndex < patternLength) {
+    if (pattern[patternIndex] == pattern[lpsLength]) {
+      lpsLength++;
+      lps[patternIndex] = lpsLength;
+      patternIndex++;
+    } else {
+      if (lpsLength != 0) {
+        lpsLength = lps[lpsLength - 1];
+      } else {
+        lps[patternIndex] = 0;
+        patternIndex++;
+      }
+    }
+  }
+}
+
+// This function checks if the string 'pattern' is an 'exact' pattern followed
+// by one or more occurences of the wildcard character, '%'; let us refer to
+// such patterns as prefix patterns. We define an 'exact' pattern as a string
+// that doesn't contain any wildcard characters; a wildcard character could
+// either be '_' or '%'. If 'pattern' is a prefix pattern, we return true.
+bool checkPrefixPattern(const std::string& pattern) {
+  size_t patternLength = pattern.size();
+  size_t patternIndex = patternLength - 1;
+
+  for (; patternIndex >= 0; patternIndex--) {
+    if (pattern[patternIndex] != '%') {
+      break;
+    }
+  }
+  if (patternIndex == patternLength - 1) {
+    return false;
+  }
+
+  for (; patternIndex >= 0; patternIndex--) {
+    if ((pattern[patternIndex] == '%') || (pattern[patternIndex] == '_')) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// This function is called when the string 'pattern' is a prefix pattern, as
+// defined in and as determined by the return value of the function bool
+// checkPrefixPattern(const std::string& pattern); When 'pattern' is a prefix
+// pattern, we match the exact pattern in the prefix pattern with the first 'm'
+// characters of the 'inputString', where m = length(pattern). If the first m
+// characters match in the two strings, we return true.
+bool matchPrefixPattern(
+    const std::string& inputString,
+    const std::string& pattern) {
+  size_t inputStringIndex = 0;
+  size_t patternIndex = 0;
+  size_t inputStringLength = inputString.size();
+  size_t patternLength = pattern.size();
+
+  for (; (patternIndex < patternLength) &&
+       (inputStringIndex < inputStringLength) && (pattern[patternIndex] != '%');
+       inputStringIndex++, patternIndex++) {
+    if (pattern[patternIndex] != inputString[inputStringIndex]) {
+      return false;
+    }
+  }
+
+  return (pattern[patternIndex] == '%');
+}
+
+// This function checks if the string 'pattern' is an 'exact' pattern preceded
+// by one or more occurences of the wildcard character, '%'; let us refer to
+// such patterns as suffix patterns. We define an 'exact' pattern as a string
+// that doesn't contain any wildcard characters; a wildcard character could
+// either be '_' or '%'. If 'pattern' is a suffix pattern, we return true.
+bool checkSuffixPattern(const std::string& pattern) {
```

Comment:
Same here too : Patterns simply % or %% will be matched by this method. Please can you check the behavior.